docs/conf.py

```diff
@@ -22,6 +22,7 @@
 # Mock imports (things that can't be installed at do building time)
 
 autodoc_mock_imports = ["pygtk", "gwy", "numpy", "pandas", "matplotlib", "seaborn", "scipy", "skimage", "pySPM"]
+import matplotlib

```

Comment:
How come this import is standard and not part of the list like above?

Reply:
The `autodoc_mock_imports` sets up a list of packages to mock if they are not installed/available when they come to be loaded.

In this instance though installing the package makes `matplotlib` available in the environment and when `grains`, `grainstats` and `run_topostats` made calls to `plot_and_save` it imported `from mpl_toolkits.axes_grid1 import make_axes_locatable` and threw this error thrice over...

```
WARNING: autodoc: failed to import module 'run_topostats' from module 'topostats'; the following exception was raised:
Traceback (most recent call last):
  File "/home/neil/.virtualenvs/topostats-docs/lib/python3.10/site-packages/sphinx/ext/autodoc/importer.py", line 58, in import_module
    return importlib.import_module(modname)
  File "/usr/lib/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1050, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1027, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1006, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 688, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 883, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/home/neil/work/projects/topostats/TopoStats/topostats/run_topostats.py", line 15, in <module>
    from topostats.grainstats import GrainStats
  File "/home/neil/work/projects/topostats/TopoStats/topostats/grainstats.py", line 15, in <module>
    from topostats.plottingfuncs import plot_and_save
  File "/home/neil/work/projects/topostats/TopoStats/topostats/plottingfuncs.py", line 8, in <module>
    from mpl_toolkits.axes_grid1 import make_axes_locatable
  File "/home/neil/.virtualenvs/topostats-docs/lib/python3.10/site-packages/mpl_toolkits/axes_grid1/__init__.py", line 5, in <module>
    from .parasite_axes import host_subplot, host_axes
  File "/home/neil/.virtualenvs/topostats-docs/lib/python3.10/site-packages/mpl_toolkits/axes_grid1/parasite_axes.py", line 161, in <module>
    with _api.suppress_matplotlib_deprecation_warning():
AttributeError: __enter__
```

The solution I hit upon yesterday was to forcibly import the mocked version of Matplotlib as above _before_ that which is in the environment is used. That seems to get round the error.

Any of the other packages in that list which are not installed in the environment but called when generating documentation using Sphinx will use the mocked version.